uefi-test-runner/build.py

```diff
@@ -162,25 +166,71 @@ def run_qemu():
     # analyzing the output of the test runner.
     ansi_escape = re.compile(r'(\x9B|\x1B\[)[0-?]*[ -/]*[@-~]')
 
-    # Start QEMU
-    qemu = sp.Popen(cmd, stdout=sp.PIPE, universal_newlines=True)
+    # Setup named pipes as a communication channel with QEMU's monitor
+    monitor_input_path = f'{qemu_monitor_pipe}.in'
+    os.mkfifo(monitor_input_path)
+    monitor_output_path = f'{qemu_monitor_pipe}.out'
+    os.mkfifo(monitor_output_path)
 
-    # Iterate over stdout...
-    for line in qemu.stdout:
-        # Strip ending and trailing whitespace + ANSI escape codes for analysis
-        stripped = ansi_escape.sub('', line.strip())
-
-        # Skip empty lines
-        if not stripped:
-            continue
-
-        # Print out the processed QEMU output to allow logging & inspection
-        print(stripped)
-
-    # Wait for QEMU to finish, then abort if that fails
-    status = qemu.wait()
-    if status != 0:
-        raise sp.CalledProcessError(cmd=cmd, returncode=status)
+    # Start QEMU
+    qemu = sp.Popen(cmd, stdin=sp.PIPE, stdout=sp.PIPE, universal_newlines=True)
+    try:
+        # Connect to the QEMU monitor
+        with open(monitor_input_path, mode='w') as monitor_input,                  \
+             open(monitor_output_path, mode='r') as monitor_output:
+            # Execute the QEMU monitor handshake, doing basic sanity checks
+            assert monitor_output.readline().startswith('{"QMP":')
+            print('{"execute": "qmp_capabilities"}', file=monitor_input, flush=True)
+            assert monitor_output.readline() == '{"return": {}}\n'
+
+            # Iterate over stdout...
+            for line in qemu.stdout:
+                # Strip ending and trailing whitespace + ANSI escape codes
+                # (This simplifies log analysis and keeps the terminal clean)
+                stripped = ansi_escape.sub('', line.strip())
+
+                # Skip lines which contain nothing else
+                if not stripped:
+                    continue
+
+                # Print out the processed QEMU output for logging & inspection
+                print(stripped)
+
+                # If the app requests a screenshot, take it
+                if stripped.startswith("SCREENSHOT: "):
```

Comment:
This if statement feels a bit large, but I'm not quite sure into which kind of abstraction I should extract it: A function? A test harness class?

Reply:
The build script would need a lot of reorganisation (and maybe rewriting it in Rust?). This approach is good enough for now